src/LayerPlan.cpp

```diff
@@ -557,6 +557,165 @@ void LayerPlan::addExtrusionMove(
+
+        if (is_overhanging != currently_overhanging_)
+        {
+            max_overhang_length_ = std::max(current_overhang_length_, max_overhang_length_);
+            current_overhang_length_ = 0;
+        }
+
+        if (is_overhanging && last_planned_position_.has_value())
+        {
+            current_overhang_length_ += vSize(target.toPoint2LL() - last_planned_position_.value());
+        }
+
+        currently_overhanging_ = is_overhanging;
+
+        addExtrusionMove(target, config, space_fill_type, flow, width_factor, spiralize, speed_factor * overhang_speed_factor, fan_speed, travel_to_z);
+    };
+
+    if (overhang_masks_.empty() || ! last_planned_position_.has_value())
+    {
+        // Unable to apply gradual overhanging (probably just disabled), just add the basic extrusion move
+        add_extrusion_move(p);
+        return;
+    }
+
+    // First, find the speed region where the segment starts
+    const Point2LL start = last_planned_position_.value();
+    size_t actual_speed_region_index = overhang_masks_.size() - 1; // Default to last region, which is infinity and beyond
+    for (const auto& [index, overhang_region] : overhang_masks_ | ranges::views::drop_last(1) | ranges::views::enumerate)
+    {
+        if (overhang_region.supported_region.inside(start, true))
+        {
+            actual_speed_region_index = index;
+            break;
+        }
+    }
+
+    // Pre-calculate the intersections of the segment with all regions (except last one, you cannot intersect an infinite plane)
+    const Point2LL end = p.toPoint2LL();
+    const Point2LL vector = end - start;
+    const coord_t segment_length = vSize(vector);
+    std::vector<std::vector<float>> speed_regions_intersections;
+    speed_regions_intersections.reserve(overhang_masks_.size() - 1);
+    for (const OverhangMask& overhang_region : overhang_masks_ | ranges::views::drop_last(1))
+    {
+        std::vector<float> intersections = overhang_region.supported_region.intersectionsWithSegment(start, end);
+        ranges::sort(intersections);
+
+        // Avoid microsegments: Move intersections that are too close to the start or end of the segment slightly more to the center of the segment.
+        for (float& intersection : intersections)
+        {
+            if (intersection * segment_length < MINIMUM_LINE_LENGTH)
+            {
+                intersection = MINIMUM_LINE_LENGTH / static_cast<float>(segment_length);
+            }
+            else if (intersection * segment_length > segment_length - MINIMUM_LINE_LENGTH)
+            {
+                intersection = (segment_length - MINIMUM_LINE_LENGTH) / static_cast<float>(segment_length);
+            }
+        }
+
+        // (Also) Avoid microsegments: Filter out pairs of intersections that are too close to each other.
+        // This should be possible because the intersections happen in the same region.
```

Comment:
I don't really understand why this should work. According to me, it shouldn't because in the next step, you always assume that you are moving from a region to an adjacent one.  But not you filter our intersections, which means you can move from a region to any other.

Reply:
I'll fix this